File: 10001~20000/17626/solution.py
import sys
import math

def solution(N):
    fourSquares = [1 for _ in range(N + 1)]
    a = int(math.sqrt(N))
    for i in range(a):
        for j in range(i*i + 1, min((i+1)*(i+1), N + 1)):
            cnt = 3
            for k in range(1, i + 1):
                tmp = fourSquares[j - k * k]
                if tmp < cnt: cnt = tmp
            fourSquares[j] = cnt + 1
    
    print(fourSquares)

    return fourSquares[N]
    

if __name__ == '__main__':
    N = int(sys.stdin.readline())

    sys.stdout.write("%d\n"%solution(N))

File: 10001~20000/17626/solution.cpp
# include <iostream>
# include <vector>
# include <cmath>

using namespace std;

int solution(int N) {
    vector<int> four_squares(N + 1);
    int a = sqrt(N);
    for (int i = 0; i < a + 1; i++){
        four_squares[i * i] = 1;
        for (int j = i*i + 1; j < (i+1)*(i+1) && j < N + 1; j++) {
            int cnt = 3;
            for (int k = 1; k <= i; k++){
                int tmp = four_squares[j - k * k];
                cnt = tmp < cnt ? tmp : cnt;
            }

            four_squares[j] = cnt + 1;
        }
    }
    
    return four_squares[N];
}

int main(void) {
    int N;
    cin >> N;
    cout << solution(N) << endl;
}